android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt

```diff
@@ -0,0 +1,208 @@
+import org.smartregister.fhircore.engine.configuration.view.RowProperties
+import org.smartregister.fhircore.engine.configuration.view.ViewProperties
+import org.smartregister.fhircore.engine.domain.model.ExtractedResource
+import org.smartregister.fhircore.engine.domain.model.RepositoryResourceData
+import org.smartregister.fhircore.engine.domain.model.ResourceData
+import org.smartregister.fhircore.engine.domain.model.RuleConfig
+import org.smartregister.fhircore.engine.domain.model.ViewType
+import org.smartregister.fhircore.engine.util.extension.extractLogicalIdUuid
+
+class RulesExecutor @Inject constructor(val rulesFactory: RulesFactory) {
+
+  suspend fun processResourceData(
+    baseResource: Resource,
+    relatedRepositoryResourceData: LinkedList<RepositoryResourceData>,
+    ruleConfigs: List<RuleConfig>
+  ): ResourceData {
+    val relatedResourcesMap = relatedRepositoryResourceData.createRelatedResourcesMap()
+    val computedValuesMap =
+      computeRules(
+        ruleConfigs = ruleConfigs,
+        baseResource = baseResource,
+        relatedResourcesMap = relatedResourcesMap
+      )
+    return ResourceData(
+      baseResourceId = baseResource.logicalId.extractLogicalIdUuid(),
+      baseResourceType = baseResource.resourceType,
+      computedValuesMap = computedValuesMap
+    )
+  }
+
+  /**
+   * This function pre-computes all the Rules for [ViewType]'s of List including list nested in the
+   * views. The LIST view computed values includes the parent's.
+   */
+  suspend fun processListResourceData(
+    listProperties: ListProperties,
+    relatedRepositoryResourceData: LinkedList<RepositoryResourceData>,
+    computedValuesMap: Map<String, Any>
+  ): List<ResourceData> {
+    val relatedResourcesMap = relatedRepositoryResourceData.createRelatedResourcesMap()
+    return listProperties.resources.flatMap { listResource ->
+      filteredListResources(relatedResourcesMap, listResource)
+        .mapToResourceData(
+          relatedResourcesMap = relatedResourcesMap,
+          ruleConfigs = listProperties.registerCard.rules,
+          listRelatedResources = listResource.relatedResources,
+          computedValuesMap = computedValuesMap
+        )
+    }
+  }
+
+  private suspend fun computeRules(
+    ruleConfigs: List<RuleConfig>,
+    baseResource: Resource,
+    relatedResourcesMap: Map<String, List<Resource>>
+  ): Map<String, Any> =
+    // Compute values via rules engine and return a map. Rule names MUST be unique
+    rulesFactory.fireRules(
+      rules = rulesFactory.generateRules(ruleConfigs),
+      baseResource = baseResource,
+      relatedResourcesMap = relatedResourcesMap
+    )
+
+  private suspend fun List<Resource>.mapToResourceData(
+    relatedResourcesMap: MutableMap<String, MutableList<Resource>>,
+    ruleConfigs: List<RuleConfig>,
+    listRelatedResources: List<ExtractedResource>,
+    computedValuesMap: Map<String, Any>
+  ) =
+    this.map { resource ->
```

Comment:
also not sure how expensive these loop iterations are, you can use parallelStream with the entrySet if that gives you a speed up